### PR TITLE
many: adding backward compatible code to upgrade SnapSetup.Flags

### PR DESCRIPTION
**Pull request by pedronis**
This adds temporary backward compatible code to upgrade SnapSetup.Flags to a new single set of shared flags in snapstate.  This is relatively straightforward because the old flag values and the snapstate ones were relatively disjunct for the flags we actually used in snapd. See flagscompat_test.go for a battery of tests to make sure it is doing the reasonable thing. 
The LeaveOld support which wasn't actually working and related GC flags have been temporarily dropped to simplify things.

Given this is about SnapSetup both the UnmarshalJSON and the special constants to keep previously used values reserved for a while can be dropped after a couple of pruning cycles (7 days), so a couple of SRUs or when we get autoupdates out since a while or 16.04.1?, if we want to be more conservative we can keep this as long as we are happy having 4 reserved potential flags we can't use.

Marked all flags related stuff in snappy package with Legacy or Interim.

### CODE REVIEW

File: daemon/api_test.go [hunk 1]
```diff
@@ -1638,10 +1638,11 @@ func (s *apiSuite) TestInstallFails(c *check.C) {
 	c.Check(chg.Err(), check.ErrorMatches, `(?sm).*Install task \(fake-install-snap-error errored\)`)
 }
 
+/* temporarely not available
```
Comment (chipaca) [inline daemon/api_test.go]: you could instead `c.Skip("temporarily not available")` which tends to not let the code bitrot too much. Also, temporarily not temporarely.

Comment (pedronis) [inline daemon/api_test.go]: it doesn't compile as is, maybe if I tweak it differently

**Comment by chipaca** [comment]
This seems fine to me. Painful, but it was going to be that, but fine. :+1:

File: overlord/ifacestate/handlers.go [hunk 2]
```diff
@@ -55,11 +55,7 @@ func (m *InterfaceManager) doSetupProfiles(task *state.Task, _ *tomb.Tomb) error
 	// Set DevMode flag if SnapSetup.Flags indicates it should be done
 	// but remember the old value in the task in case we undo.
 	task.Set("old-devmode", snapState.DevMode())
-	if ss.DevMode() {
-		snapState.Flags |= snapstate.DevMode
-	} else {
-		snapState.Flags &= ^snapstate.DevMode
-	}
+	snapState.SetDevMode(ss.DevMode())
```
Comment (mvo5) [inline overlord/ifacestate/handlers.go]: Thanks!

File: overlord/snapstate/snapmgr.go [hunk 3]
```diff
@@ -40,14 +40,38 @@ type SnapManager struct {
...
 
+type SnapSetupFlags Flags
+
+// backward compatibility: upgrade old flags based on snappy.* flags values
+// to Flags if needed
+// XXX: this can be dropped and potentially the type at the earliest
+// in 2.0.9 (after being out for about two prune cycles), at the
+// latest when we need to recover the reserved unusable flag values,
+// or this gets annoying for other reasons
+func (ssfl *SnapSetupFlags) UnmarshalJSON(b []byte) error {
```
Comment (mvo5) [inline overlord/snapstate/snapmgr.go]: Pardon my ignorance, I assume this gets implicitly tested in TestSnapSetupNewValuesUnchanged ?

**Comment by mvo5** [comment]
Looks good, thank you. This is a elegant solution for the problem I think. One question about the test, otherwise good to go as far as I'm concerned.

File: overlord/snapstate/snapmgr.go (hunk 3 above)
Comment (pedronis) [inline overlord/snapstate/snapmgr.go]: yes, it's tested by the pair of tests in flagscompat_test.go: adding these comments there for clarity

 
 func (s *flagscompatSuite) TestSnapSetupNewValuesUnchanged(c *C) {
+       // test that new snapstate flags based SnapSetup.Flags is
+       // unmarshalled as is
+

-func (s *flagscompatSuite) TestSnapSetupInterimValsUpgrade(c *C) {
+func (s *flagscompatSuite) TestSnapSetupInterimValuesUpgrade(c *C) {
+       // test that the old snappy.* flags based SnapSetup.Flags
+       // are updated correctly to the new snapstate single shared flag set
        st := state.New(nil)
        st.Lock()
        defer st.Unlock()

**Pull request merged**